### DISCUSSION TITLE
Socket.io implementation (temporary)

### DISCUSSION THREAD

**Discussion by wobsoriano** [nuxt/nuxt#15951] [Ideas]
Since we don't have access to the Nuxt instance as we did in Nuxt 2, here's a working example if you're not gonna deploy into a serverless platform:

Install socket.io and client

```
pnpm add socket.io socket.io-client
```

```ts
// server/middleware/socket.ts
import { Server } from 'socket.io'

declare global {
  var io: Server | undefined
}

export default defineEventHandler(({ node }) => {
  if (!global.io) {
    global.io = new Server(node.res.socket?.server);

    global.io.on('connection', (socket) => { })
  }
})
```

Then in your component (chat page for example):

```html
<script setup lang="ts">
import { io } from 'socket.io-client'

const connected = ref(false)

onMounted(() => {
  const socket = io();

  socket.on('connect', () => {
    connected.value = socket.connected
  });

  socket.on('disconnect', () => {
    connected.value = socket.connected
  });
})
</script>

<template>
  <div>Connected: {{ connected }}</div>
</template>
```

Made a module out of this: https://github.com/wobsoriano/nuxt3-socket.io

**Comment by sduduzog** [comment]
With the possibility of deploying to serverless platforms does this work in environments like vercel or cloudflare?

**Reply by wobsoriano** [reply]
> With the possibility of deploying to serverless platforms does this work in environments like vercel or cloudflare?

It won't work

**Comment by huangxuewu** [comment]
i was able to approach this by the nuxt3 plugin architecture:
first create 2 files under plugins folder name `socket.server.ts`  & `socket.client.ts`

in `socket.server.ts` file

import { Server } from "socket.io";
io.on("connection", async (socket) => {
    socket.emit("message","Hello world");
})

in `socket.client.ts` file

import { defineNuxtPlugin } from '#app'
import { io } from "socket.io-client";

export default defineNuxtPlugin(app => {
    const socket = io(app.$config.socket);              //access socket variable defined in nuxt.confit.ts publicRuntimeConfig
  
  socket.on("connect",()=>{
    socket.on("message",console.log)
 })
})

it is very buggy right now.

**Reply by P4sca1** [reply]
Nuxt Plugins are called on every request. So your server socket is created on every request that hits the server. This will likely not work and lead to out of memory exceptions.

**Comment by fabianwohlfart** [comment]
Both ways do not work for me. Did we progress on this already?

**Reply by wobsoriano** [reply]
What have you tried? I just tested my solution with latest version of Nuxt 3 and it still works.

Updated my answer too to include plugin based socket

**Reply by fabianwohlfart** [reply]
No you are right. I had some other errors in my app. This seems to work. Thanks.

**Comment by richardeschloss** [comment]
Hi, nuxt-socket-io addresses this problem by using Nuxt's listen hook (npm i nuxt-socket-io@next). The listen hook gives us the server instance we need if we wish to share the server with socket.io. 

If it's desired to go without nuxt-socket-io, it's still more appropriate to wait for Nuxt's listen hook before also starting up the socket.io server. And only run it once at boot (or on restarts). This can simply be done in nuxt.config:

```js
import { Server as SocketIO } from 'socket.io'
export default {
  hooks: {
    listen(server) {  
      const io = new SocketIO(server)
    }
  }
}
```

**Reply by fabianwohlfart** [reply]
At this point I must comment (for whichever folk like me digging in this topic) that the listen hook only works in dev mode at the moment. Context: https://github.com/nuxt/framework/issues/2499

**Comment by rjcpereira** [comment]
Anyone found a way to access server instance in server/api/**/* files? I'm trying to use socket-io inside the server, the client-side works using your plugin suggestion:

```javascript
//worker.server.js
export default defineNuxtPlugin(() => ({
    provide: {
        worker: async path => {
            let { data } = await useFetch(path);
            return data.value;
        }
    }
}));
```

```javascript
//worker.client.js
import { io } from "socket.io-client";

const socket = io();

const worker = async path => {
    let { data } = await useFetch(path);    
    socket.on("msg", updates => {
        console.log('msg', updates);
        if(data?.value && updates) for(let key in updates) data.value[key] = updates[key];
    });
    return data.value;
}

export default defineNuxtPlugin(() => {
    
    return {
        provide: {
            worker
        }
    };
});
```
```vue
<!--pages/dev.vue-->
<script setup>
const { $worker } = useNuxtApp();

let data = await $worker('https://fake-path.com/services/api/dev');
</script>
```

**Reply by rjcpereira** [reply]
I've created a module instead:

```javascript
import { defineNuxtModule, addServerMiddleware } from '@nuxt/kit'
import { MongoClient } from 'mongodb';
import { createClient } from 'redis';
import { Server } from 'socket.io';
import { createAdapter } from '@socket.io/redis-adapter';
import { Emitter } from '@socket.io/redis-emitter';

let io, redis, mongo;

addServerMiddleware(async (req, res, next) => {
  if(!req.url.startsWith('/api/v1/')) next();
  else {
    //stuff here
    await redis.set('xpto', 'yup');
    const data = await redis.get('xpto');
    res.end(data);
  }
});

export default defineNuxtModule({
  setup(options, nuxt) {
    
    const config = nuxt.options.privateRuntimeConfig;

    redis = createClient(config.redis);
    const socket = redis.duplicate();
    const emitter = new Emitter(redis);
    const mongodb = new MongoClient(config.mongo);

    nuxt.hook('listen', async server => {
      
      console.log('Server connected')

      io = new Server(server);
      
      io.on('connection', user => user.on('set', res => console.log(res)));
      
      Promise.all([redis.connect(), socket.connect()]).then(() => {
        console.log('Redis connected')
        io.adapter(createAdapter(redis, socket));
      }).catch(err => console.error('Error connecting redis', err))
  
      mongodb.connect().then(() => {
        mongo = mongodb.db(config.db);
        console.log('Mongo connected');
      }).catch(err => console.error('Error connecting mongo', err))
    });
  }
})```

**Reply by wobsoriano** [reply]
> please how to use `io.emit()` in `server/api`?

Updated my answer. I still prefer my original answer as I can attach the `io` to `event.context`.

**Reply by ih-hugh** [reply]
https://stackoverflow.com/questions/73592619/socket-io-websocket-not-working-in-nuxt-3-when-in-production

I tried a module-based solution and ran into issues when bundling for production.

**Comment by mei860821** [comment]
can I use this in nuxt 2 ?
I tried to reload and It still loading
<img width="102" alt="image" src="https://user-images.githubusercontent.com/72427615/169897489-4cd22998-c3a6-4220-a86d-c86782fa5620.png">
restart have error
`crbug/1173575, non-JS module files deprecated.`

I'm tried first solution
It's anyone have same question? thanks for help!

**Reply by wobsoriano** [reply]
Not sure but there's a socket.io module for nuxt2 iirc

**Comment by janfrl** [comment]
Is there a way to access the server instance inside a Nitro plugin? I want to periodically fetch data on the Nitro server from an external API and send the data to all clients. So I use a Nitro plugin to initialize the fetching on an interval. Alternative solutions would also be appreciated (but I need the data from the API on the server).

```ts
// server/plugins/fetch.ts
  export default defineNitroPlugin(async (nitroApp) => {
  setInterval((function loopFetch() {
      fetchMessages()
      $io.emit('messages', 'data')
      return loopFetch
    }()), 1000 * 60)
})
```

<details>
  <summary>I have tried the following approaches mentioned in discussion</summary>

## Middleware approach
After attaching the server instance to the event parameter, I can only access it inside `defineEventHandler` but not in `defineNitroPlugin`. I'm also unsure if it's possible to export the value somehow because all the code is inside the function, where it is not possible to use `export`.

## Nuxt plugin approach
Again, I was unable to export the server instance from `socket.server.ts` to the server folder / Nitor (in Nitro)? I tried to return it in the plugin and access it via useNuxtApp() in Nitro, but this function is not available there.

```ts
// plugins/socket.server.ts
import { Server } from 'socket.io'
import { defineNuxtPlugin } from '#app'

export default defineNuxtPlugin(({ ssrContext }) => {
  const socket = ssrContext?.event.req.socket as any
  const io = new Server(socket.server)

  io.on('connection', (socket) => {
    console.log('Made socket connection :)')

    socket.on('disconnect', () => console.log('disconnected'))
  })

  return {
    provide: {
      io,
    },
  }
})
```

```ts
// server/plugins/socket.ts
import { useNuxtApp } from '#app'

export default defineNitroPlugin(async () => {
  const { $io } = useNuxtApp() // THIS DOES NOT WORK
})
```

## Nuxt hook approach
Same problem of not being able to export it.

```ts
// nuxt.config.ts
export default defineNuxtConfig({
  hooks: {
    listen(server) {
      const io = new SocketIO(server)
      // how do I access io inside server/plugins/fetch.ts ?
    },
  }
})
```
</details>

**Reply by go4cas** [reply]
@xSuffix ... did you mange to figure out a solution? I have implemented your server plugin approach, and the socket server starts up fine, and I can access it from socket clients. I can't, however, access the `$io` from any `/server/api` endpoint. Any ideas?

**Reply by janfrl** [reply]
I have not found a solution using Sockets, but developed a workaround. In my use case I know when exactly the server refreshes its data via the external API. So instead of sending messages from the server to the clients, the server API contains the information when the client should request new data (either time in seconds or server time + next fetch time). This also makes it possible to let the frontend users know when the data is going to be updated.

**Comment by kikuchan** [comment]
The second solution doesn't work for `ssr: false` setup, just FYI.

The first solution works with node server,
but `event.context` is cleared on every request, so we need to use `globalThis` or something to preserve `io` instance.

In `server/middleware/socket.js` for example
```js
import { Server } from 'socket.io';

export default defineEventHandler((event) => {
  if (globalThis.$io) return ;
  globalThis.$io = new Server(event.req.socket?.server);


  // Example: welcome message
  $io.on('connection', (sock) => {
    sock.emit('message', 'Welcome');
  });

  // Example: send a message periodically
  let count = 0;
  setInterval(() => {
    $io.emit('message', `periodic count = ${count++}`);
  }, 1000);

  // Example: send a message with async
  (async function () {
    let count = 0;
    while (true) {
      await new Promise(resolve => setTimeout(resolve, 1000));
      $io.emit('message', `async count = ${count++}`);
    }
  })();
})
```

In `server/api/hello.js` for example
```js
// Example
export default defineEventHandler((event) => {
  $io.emit('message', "Hello");
  return 'OK';
})
```

**Reply by kikuchan** [reply]
After deep investigation, I've noticed that `event.req.socket` is *NOT* the TCP server listening on port 3000 in dev mode. (is a UNIX domain socket)
Due to the situation, transport `websocket` doesn't work in dev mode with the 1st solution. It seemingly working because it fallbacks to `polling` mode.
This is why `disconnect` event delays in dev mode.
The 1st solution works perfectly in built production mode, luckily.

**Reply by yoshevchenko** [reply]
how can i work with it in dev mode?

**Comment by mihail727** [comment]
guys please help - how to make pm2 to work with socket.io (Nuxt 3)?

**Comment by wobsoriano** [comment]
Published a module out of this

https://github.com/wobsoriano/nuxt3-socket.io

**Reply by CodeXiD** [reply]
Good day, thanks for your hard work. In my case websocket uses long pulling for some reason. Is that how it should be?

**Comment by yoshevchenko** [comment]
i try this and get xhr poll error
Has anyone experienced a similar issue or knows a solution?
